groups/bsl/bslstl/bslstl_variant.h

```diff
@@ -0,0 +1,3338 @@
+    // assignment to such an 'variant' object should not succeed.  However,
+    // unless the 'variant' object itself is 'const', it should be possible to
+    // change the value of the 'variant' object using 'emplace'.  In order to
+    // allow for that, this class manages a non-const object of 'value_type',
+    // but all the accessors return a 'const' adjusted reference to the managed
+    // object.
+
+  private:
+    // PRIVATE TYPES
+    typedef typename bsl::remove_const<TYPE>::type StoredType;
+    typedef typename bsl::allocator<char>          allocator_type;
+    typedef BloombergLP::bslmf::MovableRefUtil     MoveUtil;
+
+    // DATA
+    bsls::ObjectBuffer<StoredType> d_buffer;
+        // in-place 'TYPE' object
+
+  public:
+    // CREATORS
+
+    // MANIPULATORS
+    TYPE& value();
+        // Return the 'value_type' object in 'd_buffer' with const
+        // qualification adjusted to match that of 'TYPE'.  The behavior is
+        // undefined unless the object holds the currently active alternative.
+
+    const TYPE& value() const;
+        // Return the 'value_type' object in 'd_buffer' with const
+        // qualification adjusted to match that of 'TYPE'.  The behavior is
+        // undefined unless the object holds the currently active alternative.
+};
+
+#if !BSLS_COMPILERFEATURES_SIMULATE_CPP11_FEATURES
+template <class HEAD = void, class... ARGS>
+union VariantUnion;
+
+template <>
+union VariantUnion<void> {
+};
+
+template <class HEAD, class... TAIL>
+union VariantUnion<HEAD, TAIL...> {
+    typedef bsl::allocator<char> allocator_type;
+    Variant_DataImp<HEAD>        d_head;
+    VariantUnion<TAIL...>        d_tail;
+};
+#endif
+#endif  // BSL_VARIANT_FULL_IMPLEMENTATION
+
+                    // ===================================
+                    // struct Variant_CopyConstructVisitor
+                    // ===================================
+
+class Variant_CopyConstructVisitor {
+    // This visitor, when invoked as a non-modifiable function object on an
+    // instance of some template parameter 'TYPE', will copy-insert that
+    // instance to create an instance of the same 'TYPE' in an uninitialized
+    // buffer specified at construction of this visitor with an allocator
+    // specified at the construction of this visitor.
+
+  private:
+    // PRIVATE DATA
+    void             *d_buffer_p;
+    bslma::Allocator *d_allocator_p;
+
+  public:
+    // CREATORS
+    Variant_CopyConstructVisitor(void *buffer, bslma::Allocator *allocator)
+    : d_buffer_p(buffer)
+    , d_allocator_p(allocator)
+    {
+        BSLS_ASSERT_SAFE(d_buffer_p);
+    }
+
+    // ACCESSORS
+    template <class TYPE>
+    void operator()(const TYPE& other) const
+    {
+        typedef typename bsl::remove_cvref<TYPE>::type NEW_TYPE;
+
+        bslma::ConstructionUtil::construct(
+            reinterpret_cast<NEW_TYPE *>(d_buffer_p), d_allocator_p, other);
```

Comment:
We can do better. I've added a trivial default constructor to `VariantUnion` and `Variant_DataImp`. That makes them implicit-lifetime and takes care of any lifetime issues.
I don't know why I strip the cv qualifications here. That looks like it may be a left over from something, and it stinks of being wrong. Removed as I can't think of the reason behind it.
Thirdly, I'm modifying the internal visitors to invoke the union constructor, rather than a construction of the element. This will require them to be templated and it will add another version of the visit which adds the ID of the selected alternative. CTAD version is easy and the mock up is working. I'll do the pre-CTAD and C++03 this afternoon. Do you want the changes on a branch for easier review ?

Reply:
You don't have to put the changes in a separate branch.

Invoking the `VariantUnion` constructor seems like the right approach: it means you can explicitly initialize the desired `Variant_DataImp` member and in that case I don't think you even need implicit object creation (but there's probably something I'm missing here). BTW, now that I think about it, shouldn't `VariantUnion` be renamed to `Variant_Union` because it's component-private?

Also, is your change going to support constexpr or is that planned for later? AFAIK, for constexpr support you can't use unsigned char array buffers because reinterpret_cast isn't allowed (but in C++03 you *have* to use char array buffers because of limitations on unions) so I think it requires bifurcating the representation depending on the language version.

Reply:
You're not missing anything, I don't think you need the implicit object creation with these additional changes. However, making these unions implicit-lifetime may come in handy in other places and it's a minor change.

Will rename to `Variant_Union`.

I won't do constexpr support. The changes between C++11/C++14/C++17 with regards to constexpr make it far too complex to do under this SOW.

Reply:
I'm still seeing the union having the name `VariantUnion` instead of `Variant_Union` as it should; maybe you forgot to push a commit.

I'm concerned about the implementation of `emplace`, which is using the same `reinterpret_cast` pattern that used to be used directly here in `Variant_CopyConstructVisitor`. I don't think object creation actually makes it well-defined.

When you call the copy constructor, the default constructor of `d_union` gets called, then the visitation is performed. As the docstring says, the default constructor of `VariantUnion` leaves it with no active alternative. That means `d_head`'s lifetime has not started and `d_tail`'s lifetime has also not started.

When the visitor calls `emplace` and `emplace` performs a `reinterpret_cast` from `d_union` directly to the alternative type in order to construct an object of the alternative type, we have an issue: does this start the lifetime of the particular `Variant_DataImp` object that holds the alternative object? I think it doesn't. Sure, `Variant_DataImp` might be implicit-lifetime, but it's not enough for it to be implicit-lifetime; you also have to perform some operation that triggers implicit creation of implicit-lifetime types. I'm not seeing any such operation here.

(BTW, does `Variant_DataImp` actually need the defaulted default constructor in order to be implicit-lifetime? Wouldn't the implicitly declared copy constructor be sufficient?)

Reply:
To reiterate what we discussed yesterday, if `emplace` is implemented by recursively unwrapping the `Variant_Union` and at each step placement-newing the alternative containing the `Variant_DataImp` member that will hold the alternative being emplaced, where at the final step the `Variant_DataImp`'s constructor is called, we can avoid the need for `reinterpret_cast`, and it would resolve my concerns about lifetime UB.